File: android/app/src/main/java/social/betty/features/groupdetail/GroupDetailLogic.kt
package social.betty.features.groupdetail

import social.betty.core.model.Bet
import social.betty.core.model.Game
import social.betty.core.model.Group
import social.betty.core.model.GroupMember
import social.betty.core.model.MessageReaction
import java.time.Instant
import java.time.ZoneId
import kotlin.math.roundToInt

/**
 * Pure, derived helpers for the GroupDetail feature, ported 1:1 from the iOS
 * `GroupDetailLogic.swift` (and the web group page). Kept feature-local; the cross-platform
 * `Ranking` / `GameSchedule` live in `core/logic`.
 */

/** First own bet on a game (array order = "first wins"); a logged-out user owns none. */
fun firstOwnBet(bets: List<Bet>, gameId: Int, userId: String?): Bet? {
    if (userId == null) return null
    return bets.firstOrNull { it.gameId == gameId && it.userId == userId }
}

/** All-users bet count on a game (the bet-count chip). */
fun betCount(bets: List<Bet>, gameId: Int): Int = bets.count { it.gameId == gameId }

object GroupGameCardLogic {
    /**
     * Awarded points (below the placed-bet chip): finished games only, from the FIRST own
     * bet; null when unfinished, no own bet, the bet is unevaluated, or logged out.
     */
    fun awardedPoints(game: Game, bets: List<Bet>, userId: String?): Int? {
        if (!game.isFinished) return null
        return firstOwnBet(bets, game.id, userId)?.userPoints
    }

    /**
     * Whether the awarded-points cell should render a rocket: own bet on a finished game,
     * it's `boosted`, AND it scored > 0 points (spec §2.5 suppression rule).
     */
    fun awardedBoosted(game: Game, bets: List<Bet>, userId: String?): Boolean {
        if (!game.isFinished) return false
        val bet = firstOwnBet(bets, game.id, userId) ?: return false
        return bet.boosted && (bet.userPoints ?: 0) > 0
    }
}

/** NeedAction strip selection (web `NeedAction.vue`), scoped to a single group source. */
object NeedAction {
    sealed interface Display {
        data class Urgent(val games: List<Game>) : Display
        data class Today(val games: List<Game>) : Display
        data object Hidden : Display
    }

    /** Fractional hours until kickoff (negative when past). */
    private fun fractionalHoursUntilStart(game: Game, now: Instant): Double {
        val start = game.startDate ?: return Double.NEGATIVE_INFINITY
        return (start.epochSecond - now.epochSecond) / 3600.0
    }

    /**
     * Urgent: not finished, un-bet by me, `0 < fractional hours < 24` STRICT; sorted by
     * kickoff, capped at 3. Else today's games (any game on `now`'s calendar day). Else hidden.
     */
    fun display(
        games: List<Game>,
        bets: List<Bet>,
        userId: String?,
        now: Instant = Instant.now(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): Display {
        val sorted = games.sortedBy { it.startDate ?: Instant.EPOCH }
        val urgent = sorted.filter { game ->
            if (game.isFinished) return@filter false
            val hours = fractionalHoursUntilStart(game, now)
            if (hours <= 0.0 || hours >= 24.0) return@filter false
            firstOwnBet(bets, game.id, userId) == null
        }.take(3)
        if (urgent.isNotEmpty()) return Display.Urgent(urgent)

        val today = now.atZone(zone).toLocalDate()
        val todays = sorted.filter { game ->
            game.startDate?.atZone(zone)?.toLocalDate() == today
        }
        if (todays.isNotEmpty()) return Display.Today(todays)
        return Display.Hidden
    }
}

object GroupBetLogic {
    /** Opponents' scores show after kickoff, or any time the group allows sneak peek. */
    fun showScores(start: Instant?, peek: Boolean, now: Instant): Boolean {
        if (peek) return true
        val s = start ?: return false
        return now.isAfter(s)
    }

    fun lockInput(start: Instant?, now: Instant): Boolean {
        val s = start ?: return false
        return now.isAfter(s)
    }

    fun canSave(start: Instant?, home: String, away: String, now: Instant): Boolean {
        if (lockInput(start, now)) return false
        if (home.isEmpty() || away.isEmpty()) return false
        return true
    }

    /** Placed-bets ordering: `user_points` desc (null = 0), stable. */
    fun orderedBets(bets: List<Bet>): List<Bet> =
        bets.withIndex().sortedWith(
            compareByDescending<IndexedValue<Bet>> { it.value.userPoints ?: 0 }
                .thenBy { it.index },
        ).map { it.value }

    sealed interface SubmitRoute {
        data class Update(val betId: Int) : SubmitRoute
        data class Place(val isUniversal: Boolean) : SubmitRoute
    }

    /**
     * CRITICAL pin: only an existing bet with the all-groups box UNCHECKED routes through
     * PUT /bet/:id; every other case re-POSTs (universal edits upsert across all groups).
     */
    fun submitRoute(existing: Bet?, placeInAllGroups: Boolean): SubmitRoute =
        if (existing != null && !placeInAllGroups) {
            SubmitRoute.Update(existing.id)
        } else {
            SubmitRoute.Place(placeInAllGroups)
        }

    /** Home/tie/away distribution percentages (largest remainder, sums to exactly 100). */
    fun distribution(bets: List<Bet>): Triple<Int, Int, Int> {
        val home = bets.count { it.homeTeamScore > it.awayTeamScore }
        val away = bets.count { it.awayTeamScore > it.homeTeamScore }
        val tie = bets.count { it.homeTeamScore == it.awayTeamScore }
        return LargestRemainder.percentages(home = home, away = away, tie = tie)
    }
}

/** Largest-remainder percentages — sums to exactly 100; ties break home, away, tie. */
object LargestRemainder {
    /** Returns (home, tie, away). */
    fun percentages(home: Int, away: Int, tie: Int): Triple<Int, Int, Int> {
        val total = home + away + tie
        if (total <= 0) return Triple(0, 0, 0)
        val counts = listOf(home, away, tie) // priority: home(0), away(1), tie(2)
        val exact = counts.map { it * 100.0 / total }
        val floors = exact.map { it.toInt() }.toMutableList()
        var remaining = 100 - floors.sum()
        val order = listOf(0, 1, 2).sortedWith(
            compareByDescending<Int> { exact[it] - floors[it] }.thenBy { it },
        )
        for (index in order) {
            if (remaining <= 0) break
            floors[index] += 1
            remaining -= 1
        }
        return Triple(floors[0], floors[2], floors[1])
    }
}

/** Bet-history row result + visibility (web `UserBetListItem.vue`). */
object GroupBetRowLogic {
    enum class Result { PENDING, EXACT, WIN, MISS }

    /** `showScore = peek || processed || (start known && now > start)`. */
    fun showScore(bet: Bet, gameStart: Instant?, peek: Boolean, now: Instant): Boolean {
        if (peek) return true
        if (bet.isProcessed) return true
        if (gameStart != null && now.isAfter(gameStart)) return true
        return false
    }

    fun result(bet: Bet, showScore: Boolean, exactResultPoints: Int?): Result {
        if (!showScore || !bet.isProcessed) return Result.PENDING
        val points = bet.userPoints ?: 0
        val isExact = if (exactResultPoints != null) {
            points == exactResultPoints
        } else {
            points == 3 || points == 4
        }
        if (isExact && points > 0) return Result.EXACT
        if (points > 0) return Result.WIN
        return Result.MISS
    }
}

/**
 * Member bet-history row (web `UserHistory.vue`). A null `bet` is a "NO BET" row for a
 * game the member skipped — only included once the game has started (so we don't leak
 * who hasn't placed bets yet on upcoming games).
 */
data class UserHistoryEntry(val bet: Bet?, val game: Game)

object GroupUserHistoryLogic {
    /**
     * One row per game: bet-row if the member bet on it, skipped-row ("NO BET") if the
     * game has already started. Future games the member hasn't bet on are omitted
     * (the hidden-score / pre-kickoff pin — don't leak un-placed bets). Sorted ascending
     * by kickoff, stable.
     */
    fun entries(
        bets: List<Bet>,
        userId: String,
        games: List<Game>,
        now: Instant = Instant.now(),
    ): List<UserHistoryEntry> {
        val betByGameId = bets.asSequence()
            .filter { it.userId == userId }
            .associateBy { it.gameId }

        val rows = games.mapNotNull { game ->
            betByGameId[game.id]?.let { return@mapNotNull UserHistoryEntry(it, game) }
            val start = game.startDate
            if (start != null && now.isAfter(start)) UserHistoryEntry(null, game) else null
        }

        return rows.withIndex()
            .sortedWith(
                compareBy<IndexedValue<UserHistoryEntry>> {
                    it.value.game.startDate ?: Instant.EPOCH
                }.thenBy { it.index },
            )
            .map { it.value }
    }

    /** "<N> BETS" — only actual bets, not skipped rows. */
    fun betsCount(entries: List<UserHistoryEntry>): Int = entries.count { it.bet != null }

    /** Σ user_points over the bets only (null/skipped count as 0). */
    fun totalPoints(entries: List<UserHistoryEntry>): Int =
        entries.sumOf { it.bet?.userPoints ?: 0 }
}

object GroupStandings {
    /** Completion %: `round(complete / all * 100)`, 0 when nothing is complete. */
    fun completionPercentage(completeGames: Int, allGames: Int): Int {
        if (completeGames <= 0 || allGames <= 0) return 0
        return (completeGames.toDouble() / allGames.toDouble() * 100).roundToInt()
    }
}

/** Chat reaction grouping/toggle rules (web `toggleReaction`). */
data class ReactionGroup(val emojiId: String, val count: Int, val reactedByMe: Boolean)

object ReactionLogic {
    /** Groups reactions by emoji in first-seen order, with count + reacted-by-me flag. */
    fun grouped(reactions: List<MessageReaction>, currentUserId: String?): List<ReactionGroup> {
        val order = mutableListOf<String>()
        val counts = linkedMapOf<String, Int>()
        val mine = mutableSetOf<String>()
        for (reaction in reactions) {
            if (reaction.emojiId !in counts) order += reaction.emojiId
            counts[reaction.emojiId] = (counts[reaction.emojiId] ?: 0) + 1
            if (currentUserId != null && reaction.userId == currentUserId) mine += reaction.emojiId
        }
        return order.map { ReactionGroup(it, counts.getValue(it), it in mine) }
    }

    sealed interface ToggleAction {
        data class Set(val emojiId: String) : ToggleAction
        data object Remove : ToggleAction
    }

    /** Tap my current emoji → remove; any other → replace (one per user). null = logged out. */
    fun toggleAction(emojiId: String, reactions: List<MessageReaction>, currentUserId: String?): ToggleAction? {
        if (currentUserId == null) return null
        val mine = reactions.firstOrNull { it.userId == currentUserId }
        if (mine != null && mine.emojiId == emojiId) return ToggleAction.Remove
        return ToggleAction.Set(emojiId)
    }
}

/** Author name resolution (web `nickname || name || 'Unknown'`). */
fun chatAuthorName(member: GroupMember?): String {
    if (member == null) return "Unknown"
    member.nickname?.takeIf { it.isNotEmpty() }?.let { return it }
    member.name?.takeIf { it.isNotEmpty() }?.let { return it }
    return "Unknown"
}

/** Relative "x ago" timestamp; "just now" under a minute. */
fun relativeTime(instant: Instant?, now: Instant = Instant.now()): String {
    if (instant == null) return ""
    val seconds = now.epochSecond - instant.epochSecond
    if (seconds < 60) return "just now"
    val minutes = seconds / 60
    if (minutes < 60) return if (minutes == 1L) "1 minute ago" else "$minutes minutes ago"
    val hours = minutes / 60
    if (hours < 24) return if (hours == 1L) "1 hour ago" else "$hours hours ago"
    val days = hours / 24
    if (days < 7) return if (days == 1L) "1 day ago" else "$days days ago"
    val weeks = days / 7
    if (weeks < 5) return if (weeks == 1L) "1 week ago" else "$weeks weeks ago"
    val months = days / 30
    if (months < 12) return if (months == 1L) "1 month ago" else "$months months ago"
    val years = days / 365
    return if (years == 1L) "1 year ago" else "$years years ago"
}

/**
 * Day-group header text (web `Pools.vue`): pool names joined " & "; when the joined name
 * contains "Group" show the day title only, else "<PoolNames> - <DayTitle>".
 */
fun dayGroupHeaderText(poolNames: String, dayTitle: String): String {
    if (poolNames.isBlank()) return dayTitle
    if (poolNames.contains("Group")) return dayTitle
    return "$poolNames - $dayTitle"
}

/** Group author check (`access_level == 0`). */
fun Group.isAuthor(userId: String?): Boolean =
    userId != null && members.firstOrNull { it.userId == userId }?.accessLevel == 0

/** Display name: nickname || name || "Player". */
fun GroupMember.displayName(): String =
    nickname?.takeIf { it.isNotEmpty() }
        ?: name?.takeIf { it.isNotEmpty() }
        ?: "Player"

File: ios/Betty/Features/GroupDetail/GroupDetailLogic.swift
import Foundation

// Pure, test-pinned ports of the web group page's derived logic
// (`app/pages/dashboard/groups/[id]/index.vue`, `Pools.vue`, `Game.vue`, `BetModal.vue`,
// `UserBetListItem.vue`, `UserHistory.vue`).

// MARK: - Game date label (Game.vue `startDate`)

nonisolated enum GroupGameDateLabel {
    /// Pinned label rules:
    /// 1. finished → "Finished"
    /// 2. today AND truncated whole hours until start < 4 (includes negative, i.e. games
    ///    started earlier today past the live window) → strict relative with ceiling
    ///    rounding + clock time: "in 2 hours, 14:00" / "3 hours ago, 09:00"
    /// 3. today, >= 4h away → "Today, Fri 18:00"
    /// 4. tomorrow → "Tomorrow, Sat 15:00"
    /// 5. else → "Mon 08 Jun 12:00"
    static func text(for game: Game, at now: Date = Date(), calendar: Calendar = .current) -> String {
        if game.isFinished { return "Finished" }
        let start = game.startDate
        if calendar.isDate(start, inSameDayAs: now) {
            if GameClock.wholeHoursUntilStart(of: game, at: now) < 4 {
                return "\(strictRelative(to: start, from: now)), \(format("HH:mm", start, calendar))"
            }
            return "Today, \(format("EEE HH:mm", start, calendar))"
        }
        if let tomorrow = calendar.date(byAdding: .day, value: 1, to: now),
           calendar.isDate(start, inSameDayAs: tomorrow) {
            return "Tomorrow, \(format("EEE HH:mm", start, calendar))"
        }
        return format("EEE dd MMM HH:mm", start, calendar)
    }

    /// date-fns `formatDistanceStrict(roundingMethod: ceil, addSuffix: true)` for the
    /// seconds/minutes/hours range this label can hit (same-day distances only).
    static func strictRelative(to date: Date, from now: Date) -> String {
        let interval = date.timeIntervalSince(now)
        let isFuture = interval > 0
        let absSeconds = abs(interval)
        let phrase: String
        if absSeconds < 60 {
            let seconds = max(1, Int(absSeconds.rounded(.up)))
            phrase = seconds == 1 ? "1 second" : "\(seconds) seconds"
        } else {
            let minutes = Int((absSeconds / 60).rounded(.up))
            if minutes < 60 {
                phrase = minutes == 1 ? "1 minute" : "\(minutes) minutes"
            } else {
                let hours = Int((absSeconds / 3600).rounded(.up))
                phrase = hours == 1 ? "1 hour" : "\(hours) hours"
            }
        }
        return isFuture ? "in \(phrase)" : "\(phrase) ago"
    }

    private static func format(_ pattern: String, _ date: Date, _ calendar: Calendar) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.calendar = calendar
        formatter.timeZone = calendar.timeZone
        formatter.dateFormat = pattern
        return formatter.string(from: date)
    }
}

// MARK: - Game card derived state (Game.vue)

nonisolated enum GroupGameCardLogic {
    /// Awarded points (below the placed-bet chip): only for finished games, from the FIRST
    /// own bet in array order; hidden (nil) when unfinished, no own bet, unevaluated bet,
    /// or logged out.
    static func awardedPoints(game: Game, bets: [Bet], userID: String?) -> Int? {
        guard game.isFinished else { return nil }
        return BetOwnership.firstOwnBet(in: bets, gameID: game.id, userID: userID)?.userPoints
    }

    /// True iff the user's own bet on this finished game has `boosted == true`. The
    /// rocket-suppression check (`points > 0`) is left to the view (where the count is
    /// already in hand) so this helper doesn't have to re-derive it.
    static func awardedBoosted(game: Game, bets: [Bet], userID: String?) -> Bool {
        guard game.isFinished else { return false }
        return BetOwnership.firstOwnBet(in: bets, gameID: game.id, userID: userID)?.boosted ?? false
    }
}

// MARK: - Day-grouped schedule (Pools.vue `gameGroups`)

nonisolated struct GroupScheduleGame: Identifiable, Hashable {
    let game: Game
    let poolName: String
    var id: Int { game.id }
}

nonisolated struct GroupGameDayGroup: Identifiable, Hashable {
    /// Calendar-day key ("y-m-d").
    let key: String
    let date: Date
    /// "Today" / "Tomorrow" / "in 3 days" / "2 days ago".
    let title: String
    /// Distinct pool names in order of first appearance after the global sort.
    let poolNames: [String]
    let isNextUpcoming: Bool
    let games: [GroupScheduleGame]

    var id: String { key }

    /// Web pin: pool names joined " & "; when the joined string contains "Group" the
    /// header shows the day title only, otherwise "<PoolNames> - <DayTitle>".
    var headerText: String {
        let name = poolNames.joined(separator: " & ")
        if name.contains("Group") { return title }
        return "\(name) - \(title)"
    }
}

nonisolated enum GroupGameDaySchedule {
    static func build(pools: [PoolGames], now: Date = Date(), calendar: Calendar = .current) -> [GroupGameDayGroup] {
        // Flatten with pool-name tags, stable-sort by start date ascending.
        var flattened: [GroupScheduleGame] = []
        for pool in pools {
            flattened.append(contentsOf: pool.games.map { GroupScheduleGame(game: $0, poolName: pool.pool.name) })
        }
        let sorted = flattened.enumerated().sorted { a, b in
            if a.element.game.startDate != b.element.game.startDate {
                return a.element.game.startDate < b.element.game.startDate
            }
            return a.offset < b.offset
        }.map(\.element)

        struct Working {
            let key: String
            let date: Date
            var poolNames: [String]
            var games: [GroupScheduleGame]
            var isNextUpcoming = false
        }

        var groups: [Working] = []
        var nextUpcomingKey: String?
        for entry in sorted {
            let components = calendar.dateComponents([.year, .month, .day], from: entry.game.startDate)
            let key = "\(components.year ?? 0)-\(components.month ?? 0)-\(components.day ?? 0)"
            if let index = groups.firstIndex(where: { $0.key == key }) {
                if !groups[index].poolNames.contains(entry.poolName) {
                    groups[index].poolNames.append(entry.poolName)
                }
                groups[index].games.append(entry)
            } else {
                groups.append(Working(key: key, date: entry.game.startDate, poolNames: [entry.poolName], games: [entry]))
            }
            // First game starting at/after now flags ITS day group — pinned: if today's
            // first game already started but a later one today hasn't, today is still next.
            if nextUpcomingKey == nil, entry.game.startDate >= now {
                nextUpcomingKey = key
            }
        }

        return groups.map { working in
            GroupGameDayGroup(
                key: working.key,
                date: working.date,
                title: dayTitle(for: working.date, now: now, calendar: calendar),
                poolNames: working.poolNames,
                isNextUpcoming: working.key == nextUpcomingKey,
                games: working.games
            )
        }
    }

    static func nextUpcomingKey(in groups: [GroupGameDayGroup]) -> String? {
        groups.first(where: \.isNextUpcoming)?.key
    }

    /// date-fns `formatDistance(startOfDay(date), startOfDay(now), addSuffix: true)`
    /// for the day-distance range ("Today"/"Tomorrow" are special-cased first).
    static func dayTitle(for date: Date, now: Date, calendar: Calendar) -> String {
        if calendar.isDate(date, inSameDayAs: now) { return "Today" }
        if let tomorrow = calendar.date(byAdding: .day, value: 1, to: now),
           calendar.isDate(date, inSameDayAs: tomorrow) {
            return "Tomorrow"
        }
        let days = calendar.dateComponents(
            [.day],
            from: calendar.startOfDay(for: now),
            to: calendar.startOfDay(for: date)
        ).day ?? 0
        if days > 0 {
            return days == 1 ? "in 1 day" : "in \(days) days"
        }
        let past = abs(days)
        return past == 1 ? "1 day ago" : "\(past) days ago"
    }
}

// MARK: - Standings (group page rankedMembers / podium / champion)

nonisolated struct GroupPodiumSlot: Identifiable, Hashable {
    let place: Int
    let members: [Member]
    var id: Int { place }
}

nonisolated enum GroupStandings {
    /// Dense tie ranking on `score` (shared pinned helper).
    static func ranked(_ members: [Member]) -> [DenseRanking.Ranked<Member>] {
        DenseRanking.rank(members) { Double($0.score) }
    }

    /// Final podium: places 1–3 bucketed by place; a slot can hold multiple tied members.
    static func podium(_ ranked: [DenseRanking.Ranked<Member>]) -> [GroupPodiumSlot] {
        var slots: [GroupPodiumSlot] = []
        for entry in ranked where entry.place <= 3 {
            if let index = slots.firstIndex(where: { $0.place == entry.place }) {
                slots[index] = GroupPodiumSlot(place: entry.place, members: slots[index].members + [entry.item])
            } else {
                slots.append(GroupPodiumSlot(place: entry.place, members: [entry.item]))
            }
        }
        return slots
    }

    /// All members at place 1.
    static func champions(_ ranked: [DenseRanking.Ranked<Member>]) -> [Member] {
        ranked.filter { $0.place == 1 }.map(\.item)
    }

    static func yourPlace(_ ranked: [DenseRanking.Ranked<Member>], userID: String?) -> Int? {
        guard let userID else { return nil }
        return ranked.first { $0.item.userID == userID }?.place
    }

    /// Zero-padded place ("01") or an en dash when absent.
    static func placeDisplay(_ place: Int?) -> String {
        guard let place else { return "–" }
        return String(format: "%02d", place)
    }

    /// Completion %: `round(complete / all * 100)`, 0 when nothing is complete.
    static func completionPercentage(completeGames: Int, allGames: Int) -> Int {
        guard completeGames > 0, allGames > 0 else { return 0 }
        return Int((Double(completeGames) / Double(allGames) * 100).rounded())
    }
}

// MARK: - Bet sheet (BetModal.vue)

nonisolated enum GroupBetLogic {
    /// Sneak-peek rule: opponents' scores show after kickoff, or any time when the
    /// group allows sneak peek.
    static func showScores(start: Date, peek: Bool, now: Date = Date()) -> Bool {
        now > start || peek
    }

    static func lockInput(start: Date, now: Date = Date()) -> Bool {
        now > start
    }

    /// Save enabled only pre-kickoff with BOTH fields non-empty.
    static func canSave(start: Date, home: String, away: String, now: Date = Date()) -> Bool {
        if lockInput(start: start, now: now) { return false }
        if home.isEmpty || away.isEmpty { return false }
        return true
    }

    /// Placed-bets ordering: `user_points` descending (nil counts as 0), stable.
    static func orderedBets(_ bets: [Bet]) -> [Bet] {
        bets.enumerated().sorted { a, b in
            let pa = a.element.userPoints ?? 0
            let pb = b.element.userPoints ?? 0
            if pa != pb { return pa > pb }
            return a.offset < b.offset
        }.map(\.element)
    }

    enum SubmitRoute: Equatable {
        case update(betID: Int)
        case place(isUniversal: Bool)
    }

    /// CRITICAL regression pin: only an existing bet with the all-groups box UNCHECKED
    /// routes through PUT /bet/:id; every other case re-POSTs (universal edits must
    /// upsert across all groups of the tournament).
    static func submitRoute(existing: Bet?, placeInAllGroups: Bool) -> SubmitRoute {
        if let existing, !placeInAllGroups {
            return .update(betID: existing.id)
        }
        return .place(isUniversal: placeInAllGroups)
    }

    /// Home/tie/away distribution percentages (largest remainder, sums to exactly 100).
    static func distribution(_ bets: [Bet]) -> (home: Int, tie: Int, away: Int) {
        LargestRemainder.percentages(
            home: bets.count { $0.homeTeamScore > $0.awayTeamScore },
            away: bets.count { $0.awayTeamScore > $0.homeTeamScore },
            tie: bets.count { $0.homeTeamScore == $0.awayTeamScore }
        )
    }
}

// MARK: - Bet history row (UserBetListItem.vue)

nonisolated enum GroupBetRowLogic {
    enum Result: Equatable { case pending, exact, win, miss }

    /// `showScore = peek || processed || (game start known && now > start)`.
    static func showScore(bet: Bet, gameStart: Date?, peek: Bool, now: Date = Date()) -> Bool {
        if peek { return true }
        if bet.isProcessed { return true }
        if let gameStart, now > gameStart { return true }
        return false
    }

    /// Result styling: pending unless score is visible AND processed. Exact compares
    /// against the group's `exact_result_points` when known, else the legacy 3-or-4
    /// heuristic. Exact requires points > 0; any other positive is a win; 0 is a miss.
    static func result(bet: Bet, showScore: Bool, exactResultPoints: Int?) -> Result {
        guard showScore, bet.isProcessed else { return .pending }
        let points = bet.userPoints ?? 0
        let isExact: Bool
        if let exactResultPoints {
            isExact = points == exactResultPoints
        } else {
            isExact = points == 3 || points == 4
        }
        if isExact && points > 0 { return .exact }
        if points > 0 { return .win }
        return .miss
    }
}

// MARK: - Cover upload error copy (group page header-image flow)

nonisolated enum GroupCoverPolicy {
    static let authorOnlyMessage = "Only the group author can change the cover."
    static let unavailableMessage = "Uploads are unavailable right now. Please try again later."

    /// Cover-upload failure copy: 401/403 author gate, 503 storage outage, then the
    /// shared image-upload mapping (413 size / 415 type / generic).
    static func errorMessage(status: Int?) -> String {
        switch status {
        case 401, 403: authorOnlyMessage
        case 503: unavailableMessage
        default: ProfileImagePolicy.uploadErrorMessage(status: status)
        }
    }
}

// MARK: - Member bet history (UserHistory.vue)

/// One row in a member's bet-history sheet. `bet == nil` is a "NO BET" row for a game
/// the member skipped — only included once the game has started (so we don't leak who
/// hasn't placed bets yet on upcoming games).
nonisolated struct GroupUserHistoryEntry: Identifiable, Hashable {
    let bet: Bet?
    let game: Game
    var id: Int { game.id }
}

nonisolated enum GroupUserHistoryLogic {
    /// Build one row per game: bet-row if the member bet on it, skipped-row ("NO BET")
    /// if the game has already started. Future games the member hasn't bet on are
    /// omitted (the hidden-score / pre-kickoff pattern — don't leak un-placed bets).
    /// Sorted ascending by kickoff, stable.
    static func entries(bets: [Bet], userID: String, games: [Game], now: Date = Date()) -> [GroupUserHistoryEntry] {
        var betByGameID: [Int: Bet] = [:]
        for bet in bets where bet.userID == userID {
            betByGameID[bet.gameID] = bet
        }

        let rows: [GroupUserHistoryEntry] = games.compactMap { game in
            if let bet = betByGameID[game.id] {
                return GroupUserHistoryEntry(bet: bet, game: game)
            }
            if now > game.startDate {
                return GroupUserHistoryEntry(bet: nil, game: game)
            }
            return nil
        }

        return rows.enumerated().sorted { a, b in
            if a.element.game.startDate != b.element.game.startDate {
                return a.element.game.startDate < b.element.game.startDate
            }
            return a.offset < b.offset
        }.map(\.element)
    }

    /// "<N> BETS" — only actual bets, not skipped rows.
    static func betsCount(_ entries: [GroupUserHistoryEntry]) -> Int {
        entries.lazy.filter { $0.bet != nil }.count
    }

    /// Σ user_points over the bets only (nil/skipped count as 0).
    static func totalPoints(_ entries: [GroupUserHistoryEntry]) -> Int {
        entries.reduce(0) { $0 + ($1.bet?.userPoints ?? 0) }
    }
}
